IPv4, UDP and TCP checksum offload support

jukkar opened the pull request:
This creates infrastructure support for IPv4, UDP or TCP checksum offloading in ethernet. Currently no ethernet driver is converted to offload the checksum calculation.

jukkar committed:
tests: net: Network packet checksum offloading tests

Test that if the checksum offloading is enabled, then we do not
calculate the checksum. Also the normal case, where offloading is
disabled and we need to calculate the checksum, is tested.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka reviewed (dismissed):
This PR can be fully independent of the 2 other PR, so rebase against master.
Also: split it into:
- one patch for ethernet api
- one patch that implements it into drivers/test (though this could be already part of the first one)
- one patch that implement the net part

jukkar commented:
> This PR can be fully independent of the 2 other PR, so rebase against master.

I rather not do that as the patches are not fully independent, for example this PR fixes the VLAN tests. If I leave that change away from this set, then the corresponding change need to be introduced somewhere else, which then could cause issues with PR ordering.

jukkar committed:
net: Support network packet checksum calc offloading

Create infrastructure that allows ethernet device driver to tell
if it supports network packet checksum offloading. This applies only
to IPv4, UDP or TCP checksums. The driver can enable/disable checksum
offloading separately for Tx and Rx network packets.

If the device (ethernet in this case) can calculate the network
packet checksum for IPv4, UDP or TCP, then do not calculate the
corresponding checksum by the stack itself.

Fixes #2987

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Split the first patch to smaller pieces.

tbursztyka commented:
It's a really tiny change in vlan test: make this PR rebased on top of master. Actually this PR might go much sooner to master than the 2 others. Thus also why I want this rebase.

And you don't even need to rebase your 2 big PR on top of this one, only when it will be in master. (it's not a vital change).

jukkar commented:
> It's a really tiny change in vlan test: make this PR rebased on top of master. Actually this PR might go much sooner to master than the 2 others. Thus also why I want this rebase.
> 
> And you don't even need to rebase your 2 big PR on top of this one, only when it will be in master. (it's not a vital change).

The problem is that the change to vlan testing code is really mandatory, otherwise the test crashes. So the question is that what order these patches are applied. Currently I am not prepared to change the ordering, there are just so many patches pending that this is no longer funny.

dbkinder reviewed (approved):
+1 for docs

jukkar commented:
Regardless my earlier comment, I placed this one in front of the patch train as its functionality is later needed in VLAN PR.

jukkar commented:
sanity fixes

lpereira commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -601,9 +601,16 @@ static void eth_0_iface_init(struct net_if *iface)
+	return 0;
+}
+
+static struct ethernet_api api_funcs_0 = {
```
Shouldn't these structs be `const`?

pfalcon replied:
+1. Well, the original struct's weren't const, which may call for "make this a separate change", but I guess it's a pretty obvious change, to include into this commit.

lpereira commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -601,9 +601,16 @@ static void eth_0_iface_init(struct net_if *iface)
+
+static struct ethernet_api api_funcs_0 = {
+	.iface_api.init = eth_0_iface_init,
+	.iface_api.send = eth_tx,
```
It's often better to initialize this using more compound statements, as uninitialized elements will be set to their zero value:
```C
.iface_api = {
    .init = eth_0_iface_init,
    .send = eth_tx
```

(It might not have anything now, but if the API is expanded in the future, makes things easier to check for in `eth.h`)

pfalcon commented on drivers/ethernet/eth_dw.c:
```diff
@@ -372,9 +373,16 @@ static struct eth_runtime eth_0_runtime = {
-static struct net_if_api api_funcs = {
-	.init = eth_initialize,
-	.send = eth_tx,
+static enum eth_hw_caps eth_dw_get_capabilities(struct device *dev)
```
In https://github.com/zephyrproject-rtos/zephyr/pull/6601/commits/a20f59eea9194d7f95f8b79b7b1367f99eb089ac#diff-478140223e6b04c1b262b95c27afb803R124 , there's a code to treat caps as 0 if the function pointer is not set. Then why each and every driver is patched to include the empty function still?

jukkar replied:
> Then why each and every driver is patched to include the empty function still?

Indeed, I will remove the empty func from the drivers as that is not needed.

pfalcon reviewed (approved):
Some comments (mostly +1's for earlier comments).

Another note is that this seems to not differentiate among checksum offload types - i.e. it assumes that if there's offload, there's offload for all of IPv4 header, UDP header, TCP header. I don't know if the real hardware usually have it like that, though it makes sense to think it does.

Given that this is a split-off from a bigger PR, what I always support and ask for, let me +1 this.

jukkar commented:
> Another note is that this seems to not differentiate among checksum offload types - i.e. it assumes that if there's offload, there's offload for all of IPv4 header, UDP header, TCP header. I don't know if the real hardware usually have it like that, though it makes sense to think it does.

I have only Atmel E70 Xplained board which supports checksum offloading. In that hw one can only differentiate between RX / TX offloading but not between different protocols. If we need to add more fine grained configuration for this, we can certainly add that later.

jukkar commented:
Added fixes to native_posix ethernet board regarding checksum offloading support.

jukkar commented:
New version:

* ethernet_api in device drivers made const
* removed unused get_capabilities() functions from ethernet device drivers
